parser.y

```diff
@@ -331,19 +311,219 @@ primary_expr: ID { $$ = std::make_unique<IdExprNode>(Loc(@1), $1); }
+  }
+  /* TODO: identifier list */
+  /* The identifier may be a type name. */
+  /* TODO: direct declarator ( identifier list ) */
   ;
 
-pointer_type: type_specifier STAR { $$ = std::make_unique<PtrType>($1); }
+pointer_opt: pointer { $$ = $1; }
+  | epsilon { $$ = 0; }
   ;
 
+/* A pointer is a sequence of one or more '*'s. */
+pointer: STAR { $$ = 1; }
+  | pointer STAR { $$ = $1 + 1; }
+  ;
+
+parameter_type_list_opt: parameter_type_list { $$ = $1; }
+  | epsilon { $$ = std::vector<std::unique_ptr<ParamNode>>{}; }
+  ;
+
+parameter_type_list: parameter_list { $$ = $1; }
+  /* TODO: parameter list, ... */
+  ;
+
+parameter_list: parameter_declaration {
+    $$ = std::vector<std::unique_ptr<ParamNode>>{};
+    $$.push_back($1);
+  }
+  | parameter_list COMMA parameter_declaration {
+    auto parameter_list = $1;
+    parameter_list.push_back($3);
+    $$ = std::move(parameter_list);
+  }
+  ;
+
+parameter_declaration: declaration_specifiers declarator {
+    auto decl = $2;
+    auto resolved_type = ResolveTypeAs(std::move(decl->type), $1);
+    $$ = std::make_unique<ParamNode>(Loc(@2), std::move(decl->id), std::move(resolved_type));
+  }
+  /* Declare parameters without identifiers. */
+  | declaration_specifiers abstract_declarator_opt {
+    // TODO: for type names
+    $$ = nullptr;
+  }
+  ;
+
+abstract_declarator_opt: abstract_declarator
+  | epsilon
+  ;
+
+/* 6.7.6 Type names */
+/* NOTE: abstract means the declarator does not have an identifier */
+abstract_declarator: pointer
+  | pointer_opt direct_abstract_declarator
+  ;
+
+direct_abstract_declarator: LEFT_PAREN abstract_declarator RIGHT_PAREN
+  | direct_abstract_declarator_opt LEFT_SQUARE NUM RIGHT_SQUARE
+  | direct_abstract_declarator_opt LEFT_PAREN parameter_type_list_opt RIGHT_PAREN
+  ;
+
+direct_abstract_declarator_opt: direct_abstract_declarator
+  | epsilon
+  ;
+
+/* 6.7.8 Initialization */
+/* The current object shall have array type and the expression shall be an integer constant expression. */
+initializer: LEFT_CURLY initializer_list comma_opt RIGHT_CURLY { $$ = $2; }
+  | expr { $$ = $1; }
+  ;
+
+/* TODO: the initializer may be nested */
+initializer_list: expr {
+    $$ = std::vector<std::unique_ptr<ExprNode>>{};
+    $$.push_back($1);
+  }
+  | initializer_list COMMA expr {
+    auto initializer_list = $1;
+    initializer_list.push_back($3);
+    $$ = std::move(initializer_list);
+  }
+  ;
+
+comma_opt: COMMA
+  | epsilon
+  ;
 
 epsilon: %empty;
 %%
 
 void yy::parser::error(const yy::location& loc, const std::string& err) {
   std::cerr << loc << ": " << err << std::endl;
 }
+
+namespace {
+
+std::unique_ptr<Type> ResolveTypeAs(std::unique_ptr<Type> type,
+                                    std::unique_ptr<Type> resolved_type) {
+  // Base case: this type itself is the unknown type to resolve.
+  if (type->IsPrim()) {
+    assert(type->IsEqual(PrimitiveType::kUnknown));
```

Comment:
Like I mentioned, what if `type` is `PrimitiveType::kInt`?  Return `nullptr`?

Reply:
If the primitive type is int, it is already resolved. Calling resolve on it should indicate an internal error has occurred.